File: backend/atlas_opt/evaluator.py
"""Solver-independent evaluator for AtlasDesk expressions and functions."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from .cards import AtlasCard
from .diagnostics import Diagnostic
from .expressions import Expression, ExpressionDependency

if TYPE_CHECKING:
    from .desk import AtlasDesk


@dataclass
class EvaluationResult:
    """Structured result from evaluating an expression or function."""

    value: float | None
    diagnostics: list[Diagnostic] = field(default_factory=list)
    dependencies: list[ExpressionDependency] = field(default_factory=list)


def evaluate_expression(
    expression: Expression,
    desk: "AtlasDesk",
    card: AtlasCard | None = None,
    runtime_values: dict[str, float] | None = None,
) -> EvaluationResult:
    """Evaluate a structured expression without invoking any solver backend."""

    runtime_values = runtime_values or {}

    if expression.kind == "literal":
        return numeric_value(expression.value, "literal")

    if expression.kind == "property_ref":
        if card is None:
            return EvaluationResult(
                None,
                [Diagnostic("error", "Property reference requires a card context.")],
                expression.dependencies(None),
            )
        dependency = expression.dependencies(card.id)
        property_name = expression.property_name or ""
        runtime_key = f"{card.id}.{property_name}"
        if runtime_key in runtime_values:
            return EvaluationResult(runtime_values[runtime_key], dependencies=dependency)
        prop = card.property_by_name(property_name)
        if prop is None:
            return EvaluationResult(
                None,
                [
                    Diagnostic(
                        "warning",
                        f'Card "{card.id}" is missing property "{property_name}".',
                        card.id,
                    )
                ],
                dependency,
            )
        if prop.kind == "decision_ref" and isinstance(prop.value, str):
            if prop.value in runtime_values:
                return EvaluationResult(runtime_values[prop.value], dependencies=dependency)
            dynamic_key = f"{card.id}.{property_name}"
            if dynamic_key in runtime_values:
                return EvaluationResult(runtime_values[dynamic_key], dependencies=dependency)
        if prop.kind == "data_ref":
            return evaluate_data_reference(prop.value, desk, f"{card.id}.{property_name}", dependency)
        value = numeric_value(prop.value, f'{card.id}.{property_name}')
        value.dependencies = dependency
        return value

    if expression.kind == "multiply":
        left = evaluate_required_child(expression.left, desk, card, runtime_values, "left")
        right = evaluate_required_child(expression.right, desk, card, runtime_values, "right")
        return combine_binary(left, right, "multiply")

    if expression.kind == "add":
        results = [
            evaluate_expression(term, desk, card, runtime_values) for term in expression.terms
        ]
        diagnostics = [diag for result in results for diag in result.diagnostics]
        dependencies = [dep for result in results for dep in result.dependencies]
        if any(result.value is None for result in results):
            return EvaluationResult(None, diagnostics, dependencies)
        return EvaluationResult(sum(result.value or 0 for result in results), diagnostics, dependencies)

    if expression.kind == "function_ref":
        if not expression.function_card_id:
            return EvaluationResult(
                None,
                [Diagnostic("warning", "Function reference is missing a function card id.")],
                expression.dependencies(),
            )
        result = desk.evaluate_function(expression.function_card_id, runtime_values=runtime_values)
        return EvaluationResult(
            result.value,
            result.diagnostics,
            [*expression.dependencies(), *result.dependencies],
        )

    return EvaluationResult(
        None,
        [Diagnostic("error", f'Unsupported expression kind "{expression.kind}".')],
        [],
    )


def evaluate_required_child(
    expression: Expression | None,
    desk: "AtlasDesk",
    card: AtlasCard | None,
    runtime_values: dict[str, float],
    label: str,
) -> EvaluationResult:
    """Evaluate a required child expression, returning a diagnostic if absent."""

    if expression is None:
        return EvaluationResult(None, [Diagnostic("error", f"Missing {label} expression.")], [])
    return evaluate_expression(expression, desk, card, runtime_values)


def combine_binary(left: EvaluationResult, right: EvaluationResult, operator: str) -> EvaluationResult:
    """Combine two numeric results."""

    diagnostics = [*left.diagnostics, *right.diagnostics]
    dependencies = [*left.dependencies, *right.dependencies]
    if left.value is None or right.value is None:
        return EvaluationResult(None, diagnostics, dependencies)
    if operator == "multiply":
        return EvaluationResult(left.value * right.value, diagnostics, dependencies)
    return EvaluationResult(None, [*diagnostics, Diagnostic("error", f"Unsupported operator {operator}.")], dependencies)


def numeric_value(value: object, label: str) -> EvaluationResult:
    """Coerce a value to float or return a non-crashing diagnostic."""

    if isinstance(value, bool) or value is None:
        return EvaluationResult(None, [Diagnostic("warning", f"{label} is not numeric.")], [])
    if isinstance(value, (int, float)):
        return EvaluationResult(float(value), [], [])
    if isinstance(value, str) and value.strip():
        try:
            return EvaluationResult(float(value), [], [])
        except ValueError:
            pass
    return EvaluationResult(None, [Diagnostic("warning", f"{label} is not numeric.")], [])


def evaluate_data_reference(value: object, desk: "AtlasDesk", label: str, dependency) -> EvaluationResult:
    """Resolve a tiny CSV data reference from Data-card preview metadata."""

    if not isinstance(value, dict):
        return EvaluationResult(None, [Diagnostic("warning", f"{label} data_ref is not structured.")], dependency)
    data_card_id = value.get("dataCardId")
    column = value.get("column")
    if not isinstance(data_card_id, str) or not isinstance(column, str):
        return EvaluationResult(None, [Diagnostic("warning", f"{label} data_ref requires dataCardId and column.")], dependency)
    data_card = desk.cards.get(data_card_id)
    if data_card is None or data_card.source is None or data_card.source.data is None:
        return EvaluationResult(None, [Diagnostic("warning", f'Data card "{data_card_id}" was not found.', data_card_id)], dependency)
    if column not in data_card.source.data.columns:
        return EvaluationResult(None, [Diagnostic("warning", f'Column "{column}" was not found in "{data_card_id}".', data_card_id)], dependency)
    row_index = value.get("rowIndex", 0)
    if not isinstance(row_index, int):
        row_index = 0
    try:
        row = data_card.source.data.previewRows[row_index]
    except IndexError:
        return EvaluationResult(None, [Diagnostic("warning", f"{label} rowIndex is outside preview data.")], dependency)
    result = numeric_value(row.get(column), f"{data_card_id}.{column}")
    result.dependencies = dependency
    return result

File: apps/web/src/atlas/core/evaluator.ts
import { createConstraintConfig } from "./constraints";
import { evaluateAtlasQuery } from "./queries";
import { createObjectiveConfig } from "./objectives";
import type {
  AtlasCard,
  AtlasConstraintExpression,
  AtlasExpression,
  AtlasWorkbenchState
} from "./types";

export type AtlasEvaluationDiagnostic = {
  level: "warning" | "error";
  message: string;
  cardId?: string;
  propertyName?: string;
};

export type AtlasEvaluationValue =
  | { kind: "number"; value: number }
  | { kind: "constraint"; left: number | null; right: number | null; satisfied: boolean | null };

export type AtlasEvaluationEntry = {
  cardId: string;
  label: string;
  value: AtlasEvaluationValue | null;
  diagnostics: AtlasEvaluationDiagnostic[];
};

export type AtlasEvaluationReport = {
  entries: Record<string, AtlasEvaluationEntry>;
  diagnostics: AtlasEvaluationDiagnostic[];
  mode?: AtlasEvaluationMode;
};

export type AtlasEvaluationMode = "current" | "solution";

export type AtlasEvaluationOptions = {
  mode?: AtlasEvaluationMode;
  runtimeValues?: Record<string, number>;
};

type EvaluationContext = {
  state: AtlasWorkbenchState;
  entries: Record<string, AtlasEvaluationEntry>;
  evaluating: Set<string>;
  runtimeValues: Record<string, number>;
  mode: AtlasEvaluationMode;
};

type NumericResult = {
  value: number | null;
  diagnostics: AtlasEvaluationDiagnostic[];
};

export function evaluateAtlasWorkbench(
  state: AtlasWorkbenchState,
  options: AtlasEvaluationOptions = {}
): AtlasEvaluationReport {
  const context: EvaluationContext = {
    state,
    entries: {},
    evaluating: new Set(),
    runtimeValues: options.runtimeValues ?? {},
    mode: options.mode ?? "current"
  };

  for (const card of state.cards) {
    evaluateAtlasCard(card, context);
  }

  return {
    entries: context.entries,
    diagnostics: Object.values(context.entries).flatMap((entry) => entry.diagnostics),
    mode: context.mode
  };
}

export function evaluateAtlasCard(
  card: AtlasCard,
  contextOrState: EvaluationContext | AtlasWorkbenchState
): AtlasEvaluationEntry {
  const context = isEvaluationContext(contextOrState)
    ? contextOrState
    : {
        state: contextOrState,
        entries: {},
        evaluating: new Set<string>(),
        runtimeValues: {},
        mode: "current" as const
      };
  const cached = context.entries[card.id];
  if (cached) return cached;

  if (context.evaluating.has(card.id)) {
    return makeEntry(card, null, [
      { level: "error", cardId: card.id, message: `Circular evaluation at ${card.title}.` }
    ]);
  }

  context.evaluating.add(card.id);
  const entry =
    card.type === "function"
      ? evaluateFunctionCard(card, context)
      : card.type === "objective"
        ? evaluateObjectiveCard(card, context)
        : card.type === "constraint"
          ? evaluateConstraintCard(card, context)
          : makeEntry(card, null, []);
  context.entries[card.id] = entry;
  context.evaluating.delete(card.id);
  return entry;
}

export function getEvaluationEntry(
  report: AtlasEvaluationReport | null,
  cardId: string | null | undefined
) {
  return cardId && report ? report.entries[cardId] ?? null : null;
}

function evaluateFunctionCard(card: AtlasCard, context: EvaluationContext): AtlasEvaluationEntry {
  if (card.functionKind !== "tagged_sum") return makeEntry(card, null, []);
  const queryId = card.taggedSum?.queryId;
  const expression = card.taggedSum?.expression;
  const query = queryId ? context.state.queries.find((candidate) => candidate.id === queryId) : null;

  if (!query || !expression) {
    return makeEntry(card, null, [
      {
        level: "warning",
        cardId: card.id,
        message: "TaggedSum needs a query and expression before it can evaluate."
      }
    ]);
  }

  const matchedCards = evaluateAtlasQuery(query, context.state.cards);
  const diagnostics: AtlasEvaluationDiagnostic[] = [];
  let total = 0;

  for (const matchedCard of matchedCards) {
    const result = evaluateExpression(expression, matchedCard, context.runtimeValues);
    diagnostics.push(...result.diagnostics);
    if (result.value === null) continue;
    total += result.value;
  }

  return makeEntry(
    card,
    diagnostics.some((diagnostic) => diagnostic.level === "error")
      ? null
      : { kind: "number", value: total },
    diagnostics
  );
}

function evaluateObjectiveCard(card: AtlasCard, context: EvaluationContext): AtlasEvaluationEntry {
  const objective = createObjectiveConfig(card.objective);
  const diagnostics: AtlasEvaluationDiagnostic[] = [];
  let total = 0;

  for (const term of objective.terms) {
    const functionCard = term.functionCardId
      ? context.state.cards.find((candidate) => candidate.id === term.functionCardId)
      : null;
    if (!functionCard) {
      diagnostics.push({
        level: "warning",
        cardId: card.id,
        message: `Objective term "${term.name}" does not reference a Function card.`
      });
      continue;
    }

    const entry = evaluateAtlasCard(functionCard, context);
    diagnostics.push(...entry.diagnostics);
    const value = numericEntryValue(entry);
    if (value === null) {
      diagnostics.push({
        level: "error",
        cardId: card.id,
        message: `Objective term "${term.name}" could not evaluate ${functionCard.title}.`
      });
      continue;
    }
    total += value;
  }

  return makeEntry(
    card,
    diagnostics.some((diagnostic) => diagnostic.level === "error")
      ? null
      : { kind: "number", value: total },
    diagnostics
  );
}

function evaluateConstraintCard(card: AtlasCard, context: EvaluationContext): AtlasEvaluationEntry {
  const constraint = createConstraintConfig(card.constraint);
  const left = evaluateConstraintExpression(constraint.left, card, context);
  const right = evaluateConstraintExpression(constraint.right, card, context);
  const diagnostics = [...left.diagnostics, ...right.diagnostics];
  const satisfied =
    left.value === null || right.value === null
      ? null
      : constraint.operator === "<="
        ? left.value <= right.value
        : constraint.operator === ">="
          ? left.value >= right.value
          : left.value === right.value;

  return makeEntry(
    card,
    {
      kind: "constraint",
      left: left.value,
      right: right.value,
      satisfied
    },
    diagnostics
  );
}

function evaluateConstraintExpression(
  expression: AtlasConstraintExpression,
  owner: AtlasCard,
  context: EvaluationContext
): NumericResult {
  if (expression.kind === "constant") return { value: expression.value, diagnostics: [] };
  const functionCard = expression.functionCardId
    ? context.state.cards.find((card) => card.id === expression.functionCardId)
    : null;
  if (!functionCard) {
    return {
      value: null,
      diagnostics: [
        {
          level: "warning",
          cardId: owner.id,
          message: "Constraint expression does not reference a Function card."
        }
      ]
    };
  }

  const entry = evaluateAtlasCard(functionCard, context);
  return { value: numericEntryValue(entry), diagnostics: entry.diagnostics };
}

export function evaluateExpression(
  expression: AtlasExpression,
  card: AtlasCard,
  runtimeValues: Record<string, number> = {}
): NumericResult {
  switch (expression.kind) {
    case "literal":
      return parseNumericValue(expression.value, {
        level: "error",
        cardId: card.id,
        message: `Literal "${expression.value}" is not numeric.`
      });
    case "property_ref":
      return evaluatePropertyReference(card, expression.propertyName, runtimeValues);
    case "multiply": {
      const left = evaluateExpression(expression.left, card, runtimeValues);
      const right = evaluateExpression(expression.right, card, runtimeValues);
      return {
        value: left.value === null || right.value === null ? null : left.value * right.value,
        diagnostics: [...left.diagnostics, ...right.diagnostics]
      };
    }
    case "add": {
      const results = expression.terms.map((term) => evaluateExpression(term, card, runtimeValues));
      const diagnostics = results.flatMap((result) => result.diagnostics);
      if (results.some((result) => result.value === null)) return { value: null, diagnostics };
      return {
        value: results.reduce((sum, result) => sum + (result.value ?? 0), 0),
        diagnostics
      };
    }
    default:
      return {
        value: null,
        diagnostics: [
          {
            level: "error",
            cardId: card.id,
            message: "Expression kind is not supported by the prototype evaluator."
          }
        ]
      };
  }
}

function evaluatePropertyReference(
  card: AtlasCard,
  propertyName: string,
  runtimeValues: Record<string, number>
): NumericResult {
  const property = card.properties.find((candidate) => candidate.name === propertyName);
  if (!property) {
    return {
      value: null,
      diagnostics: [
        {
          level: "error",
          cardId: card.id,
          propertyName,
          message: `${card.title} is missing property "${propertyName}".`
        }
      ]
    };
  }

  if (property.kind === "decision_ref") {
    const decisionId = typeof property.value === "string" ? property.value : "";
    const runtimeKey = decisionId || `${card.id}.${property.name}`;
    if (typeof runtimeValues[runtimeKey] === "number") {
      return { value: runtimeValues[runtimeKey], diagnostics: [] };
    }
    if (typeof runtimeValues[`${card.id}.${property.name}`] === "number") {
      return { value: runtimeValues[`${card.id}.${property.name}`], diagnostics: [] };
    }
  }

  if (property.kind === "data_ref" && typeof property.value === "object" && property.value !== null) {
    return {
      value: null,
      diagnostics: [
        {
          level: "warning",
          cardId: card.id,
          propertyName,
          message: `${card.title}.${propertyName} references ${property.value.dataCardId}.${property.value.column}; backend evaluation resolves CSV data references.`
        }
      ]
    };
  }

  return parseNumericValue(
    typeof property.value === "object" ? null : property.value,
    {
    level: "error",
    cardId: card.id,
    propertyName,
    message: `${card.title}.${propertyName} does not have a numeric value.`
    }
  );
}

function parseNumericValue(
  value: string | number | boolean | null,
  diagnostic: AtlasEvaluationDiagnostic
): NumericResult {
  if (typeof value === "number" && Number.isFinite(value)) return { value, diagnostics: [] };
  if (typeof value === "string" && value.trim()) {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return { value: parsed, diagnostics: [] };
  }
  return { value: null, diagnostics: [diagnostic] };
}

function numericEntryValue(entry: AtlasEvaluationEntry) {
  return entry.value?.kind === "number" ? entry.value.value : null;
}

function makeEntry(
  card: AtlasCard,
  value: AtlasEvaluationValue | null,
  diagnostics: AtlasEvaluationDiagnostic[]
): AtlasEvaluationEntry {
  return {
    cardId: card.id,
    label: card.title,
    value,
    diagnostics
  };
}

function isEvaluationContext(value: EvaluationContext | AtlasWorkbenchState): value is EvaluationContext {
  return "state" in value && "entries" in value;
}
